cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -1425,31 +1431,86 @@ class AsofJoinNode : public ExecNode {
     // The reason for this is that there are cases at the end of a table where we don't
     // know whether the RHS of the join is up-to-date until we know that the table is
     // finished.
-    process_.Push(true);
+    PushProcess(true);
+
     return Status::OK();
   }
+  void PushProcess(bool value) {
+#ifdef ARROW_ENABLE_THREADING
+    process_.Push(value);
+#else
+    if (value) {
+      ProcessNonThreaded();
+    }
```

Comment:
I notice in the sorted merge node you have an else branch here which calls `EndFromSingleThread`.  Is that needed here?

Reply:
I have the same concern. For the threaded version, a poison pill will effectively lead to a cleanup (`EndFromProcessThread`) in both asof join (in `ProcessThread`) and sorted merge join (in `EmitBatches`). I think this is necessary for cases when upstream calls `StopProducing` which takes a poison pill.

Reply:
I'm not sure, because there isn't really a producer thread here, but it doesn't do anything bad adding it in, so I guess it is safer.